File: src/cache/pollution-guard.js
/**
 * Cache Pollution Defense Engine
 * Detects low-reuse flood patterns and protects high-value cached items from eviction churn.
 */

import { POLLUTION_RISK } from '../core/types.js';

export class PollutionGuard {
  constructor() {
    this.windowRequests = [];
    this.windowMaxDuration = 10; // 10 seconds sliding window
    this.riskLevel = POLLUTION_RISK.LOW;
    this.uniqueKeyRate = 0.12;
    this.usefulOccupancyPercent = 85;
    this.protectedKeys = new Set();
    this.quarantinedKeys = new Set();
  }

  recordRequest(key, isPollutionKey, timestamp) {
    this.windowRequests.push({ key, isPollutionKey: !!isPollutionKey, timestamp });
    
    // Purge old requests outside window
    const cutoff = timestamp - this.windowMaxDuration;
    while (this.windowRequests.length > 0 && this.windowRequests[0].timestamp < cutoff) {
      this.windowRequests.shift();
    }
  }

  evaluate(cacheEntries, currentTime) {
    const totalInWindow = this.windowRequests.length;
    if (totalInWindow < 10) {
      this.riskLevel = POLLUTION_RISK.LOW;
      this.uniqueKeyRate = 0.15;
      return this.getStatus();
    }

    const uniqueKeys = new Set(this.windowRequests.map(r => r.key));
    const uniqueRatio = uniqueKeys.size / totalInWindow;
    this.uniqueKeyRate = Number(uniqueRatio.toFixed(3));

    // Determine Pollution Risk Level
    if (uniqueRatio > 0.70 || this.windowRequests.filter(r => r.isPollutionKey).length / totalInWindow > 0.40) {
      this.riskLevel = POLLUTION_RISK.HIGH;
    } else if (uniqueRatio > 0.40) {
      this.riskLevel = POLLUTION_RISK.MEDIUM;
    } else {
      this.riskLevel = POLLUTION_RISK.LOW;
    }

    // Evaluate useful cache occupancy vs single-hit wasted space
    let totalBytes = 0;
    let usefulBytes = 0;
    this.protectedKeys.clear();

    for (const entry of cacheEntries) {
      totalBytes += entry.sizeBytes;
      const isHighValue = (entry.score >= 0.55 || entry.totalHits >= 3);
      if (isHighValue) {
        usefulBytes += entry.sizeBytes;
        if (this.riskLevel !== POLLUTION_RISK.LOW) {
          this.protectedKeys.add(entry.id);
        }
      }
    }

    this.usefulOccupancyPercent = totalBytes > 0 
      ? Math.round((usefulBytes / totalBytes) * 100)
      : 80;

    return this.getStatus();
  }

  isProtected(key) {
    return this.riskLevel !== POLLUTION_RISK.LOW && this.protectedKeys.has(key);
  }

  getStatus() {
    let explanation = "Normal traffic access patterns. Useful cache occupancy is optimal.";
    if (this.riskLevel === POLLUTION_RISK.HIGH) {
      explanation = "CRITICAL: Low-reuse objects are flooding the cache. High-value resident objects are actively protected against eviction.";
    } else if (this.riskLevel === POLLUTION_RISK.MEDIUM) {
      explanation = "WARNING: Elevated unique-key rate detected. Probationary admission threshold enabled.";
    }

    return {
      riskLevel: this.riskLevel,
      uniqueKeyRatePercent: Math.round(this.uniqueKeyRate * 100),
      usefulOccupancyPercent: this.usefulOccupancyPercent,
      protectedCount: this.protectedKeys.size,
      explanation,
      isMitigating: this.riskLevel !== POLLUTION_RISK.LOW
    };
  }

  reset() {
    this.windowRequests = [];
    this.riskLevel = POLLUTION_RISK.LOW;
    this.uniqueKeyRate = 0.12;
    this.usefulOccupancyPercent = 85;
    this.protectedKeys.clear();
    this.quarantinedKeys.clear();
  }
}

File: backend/cache/pollution_guard.py
"""
Cache Pollution Defense System
Detects unique key deluges (crawlers, scans, scrapers) and shields high-value resident objects from cache thrashing.
"""

from collections import deque
from typing import Dict, Any, List, Set
from backend.core.types import PollutionRisk


class PollutionGuard:
    def __init__(self, window_size: int = 150):
        self.window_size = window_size
        self.request_history: deque = deque(maxlen=window_size)
        self.key_counts: Dict[str, int] = {}
        self.risk_level = PollutionRisk.LOW.value
        self.unique_key_rate = 0.0
        self.useful_occupancy = 85.0
        self.protected_keys: Set[str] = set()

    def record_request(self, key: str, is_pollution_key: bool = False, current_time: float = 0.0):
        if len(self.request_history) == self.window_size:
            oldest_key = self.request_history[0]
            if oldest_key in self.key_counts:
                self.key_counts[oldest_key] -= 1
                if self.key_counts[oldest_key] <= 0:
                    del self.key_counts[oldest_key]

        self.request_history.append(key)
        self.key_counts[key] = self.key_counts.get(key, 0) + 1

    def evaluate(self, resident_entries: List[Any], current_time: float) -> Dict[str, Any]:
        total_recent = len(self.request_history)
        if total_recent == 0:
            return self.get_status()

        unique_keys = len(self.key_counts)
        self.unique_key_rate = unique_keys / float(total_recent)

        # Risk Classification
        if self.unique_key_rate > 0.65:
            self.risk_level = PollutionRisk.HIGH.value
        elif self.unique_key_rate > 0.40:
            self.risk_level = PollutionRisk.MEDIUM.value
        else:
            self.risk_level = PollutionRisk.LOW.value

        # Calculate useful occupancy & select shielded items
        self.protected_keys.clear()
        useful_count = 0
        
        for entry in resident_entries:
            score = getattr(entry, "score", 0.5)
            hits = getattr(entry, "total_hits", 1)
            
            # High utility items get protected
            if score >= 0.50 or hits >= 3:
                useful_count += 1
                if self.risk_level in [PollutionRisk.HIGH.value, PollutionRisk.MEDIUM.value]:
                    self.protected_keys.add(getattr(entry, "id", ""))
            elif score >= 0.35:
                useful_count += 1

        total_resident = max(1, len(resident_entries))
        self.useful_occupancy = (useful_count / float(total_resident)) * 100.0

        return self.get_status()

    def is_protected(self, key: str) -> bool:
        return key in self.protected_keys

    def get_status(self) -> Dict[str, Any]:
        return {
            "riskLevel": self.risk_level,
            "uniqueKeyRatePercent": round(self.unique_key_rate * 100.0, 1),
            "usefulOccupancyPercent": round(self.useful_occupancy, 1),
            "protectedItemsCount": len(self.protected_keys),
            "windowRequests": len(self.request_history)
        }
